utility/src/main/java/org/oppia/android/util/locale/AndroidLocaleFactory.kt

```diff
@@ -55,13 +55,26 @@ class AndroidLocaleFactory @Inject constructor(
    */
   fun createAndroidLocale(localeContext: OppiaLocaleContext): Locale {
     // Note: computeIfAbsent is used here instead of getOrPut to ensure atomicity across multiple
-    // threads calling into this create function.
-    return memoizedLocales.computeIfAbsent(localeContext) {
```

Comment:
Given the problems that this has caused in earlier Android API versions, perhaps we should outright prohibit it to ensure this can't happen again.

I suggest adding a static regex pattern + test to check for this. See: https://github.com/oppia/oppia-android/wiki/Static-Analysis-Checks#generic-regex-pattern-matching-against-file-contents.

Reply:
Thanks @BenHenning 
Checked the suggested approach : [](https://gist.github.com/BenHenning/deb77757e89e1974066a6e76d1496f8b/revisions)

I have conducted a test on kotlin playground with suggested approach using this docs : [](https://kotlinlang.org/docs/shared-mutable-state-and-concurrency.html)

But synchronization doesn't seem to be achieved . You could see the test here  :  [](https://pl.kotl.in/cgtBwrYAb) 
I think alternatively we can consider using mutex ,  Mutex Test : [](https://pl.kotl.in/Q9BFy_zak)
Mutex runs faster than blockingDispatcher from the tests. 
We may also need to consider what the docs have to say about using Mutex : 

> The locking in this example is fine-grained, so it pays the price

What do you suggest ?

Reply:
@BenHenning PTAL

Reply:
Thanks @Shaik-Sirajuddin.

I think we would need to run the checks in a carefully isolated environment (vs. a cloud environment) in order to be confident in the results, but it does make sense why a fine-grained confinement strategy like blocking dispatcher would have poorer performance over a mutex (especially a cooperative one like Kotlin's ``Mutex``--I actually wasn't aware of that one).

I generally lean toward maintaining the existing pattern for properly solving this (which is blocking dispatcher as we use that everywhere else). I'm actually not opposed to using ``Mutex``, but one concern that comes up is how easy it is to deadlock when using managed critical sections (though this problem isn't completely fixed by blocking dispatcher; it's possible to live-lock that in the right circumstances). Generally the best solution is to use a message queue (like Kotlin's actor which we use in highly complex state machines in the app).

I'm open to either ``Mutex`` or blocking dispatcher here since both will play well with coroutines.